Homework10.2

jfdsmit opened the pull request.

jfdsmit committed:
semblance of a LinkedList

jfdsmit committed:
a linked list container

jfdsmit committed:
unit test for const_iterator

jfdsmit committed:
dereference test and const_iterator deref test

maciekgajewski commented on course5/homework10.2/linked_list.h:
```diff
@@ -0,0 +1,148 @@
+	struct Node
+	{
+		T value;
+		std::unique_ptr<Node> next{nullptr};
```
nullptr not needed

Commit:
avoiding some default constructors

maciekgajewski commented on course5/homework10.2/linked_list.h:
```diff
@@ -0,0 +1,148 @@
+			current = current->next.get();
+			return old;
+		}
+		T const & operator*() const { return current->value; }
```
not sure about this..
Should you have
```
T& operator*();
const T& operator*() const;
```
?

maciekgajewski commented on course5/homework10.2/linked_list.h:
```diff
@@ -0,0 +1,148 @@
+		bool operator!=(const_iterator const & that) const { return this->current != that.current; }
+	private:
+		Node const * current{nullptr};
+	};
```
```operator->``` is missing

maciekgajewski commented on course5/homework10.2/linked_list.h:
```diff
@@ -0,0 +1,148 @@
+	template <class InputIt>
+	LinkedList(InputIt const & first, InputIt const & last)
+	{
+		std::unique_ptr<Node> * current{&head};
```
Why are you using this curly-bracket initialisation everywhere? Makes code harder to read.

maciekgajewski commented on course5/homework10.2/linked_list.h:
```diff
@@ -0,0 +1,148 @@
+	LinkedList & operator=(LinkedList const & that)
+	{
+		this->head = nullptr;
+		std::unique_ptr<Node> * current{&head};
```
This is the same non-trivial code as in the function above. Consider putting it into common function.

maciekgajewski commented on course5/homework10.2/linked_list.h:
```diff
@@ -0,0 +1,148 @@
+
+	bool operator==(LinkedList const & that) const
+	{
+		auto itl = begin(), itr = that.begin();
```
Consider using ```std::equal```

maciekgajewski commented on course5/homework10.2/linked_list.h:
```diff
@@ -0,0 +1,148 @@
+		head = std::move(newNode);
+	}
+
+	void pop_front() { if (head) head = std::move(head->next); }
```
The ```if``` breaks "don't pay for what you're not using" principle.

jfdsmit replied:
What's the best solution? Document that popping from an empty list is undefined and let the program crash?

maciekgajewski commented on course5/homework10.2/linked_list.h:
```diff
@@ -0,0 +1,148 @@
+	std::size_t size()
+	{
+		std::size_t result{0};
+		for (auto it = begin(); it != end(); ++it)
```
Consider using ```std::distance```

maciekgajewski reviewed (commented):
Nice!
But your list has one flaw: it requires T to be default-constructible.

Try creating a type without default constructor, and storing it in your list.

Commit:
test coverage for not-default constructible type

jfdsmit commented:
Yeah, I realized the default construction problem on the train, tried to fix it on the train.

Commit:
Processed code review
- removed unnecessary nullptr inits
- operator-> on iterators
- unified code paths for copy constructor and construction from range
- std::equal and std::distance for operator== and size()
